### PR TITLE
[feat] PL mvp1: validation + logging

### PR DESCRIPTION
**Pull request by ytsheng**
- [x] MVP 1. Evaluation and Logging - Eval accuracy parity between mmf_trainer and lightning_trainer
  -  [X] Validation at the right frequency and at the end (also calculate the right metrics)
  -  [X] Do logging (same logging format with slight difference is ok)
- [ ] Tests - tests only done for pytorch lightning integration
  - [ ] test logging values are the same as `mmf_trainer`'s logging values
  - [ ] test eval values are the same as `mmf_trainer`'s eval values
  - [ ] test tensorboard logging the same as `mmf_trainer`'s

### CODE REVIEW

File: mmf/trainers/lightning_core/loop_callback.py [hunk 1]
```diff
@@ -2,35 +2,196 @@
...
     def __init__(self, lightning_trainer):
         super().__init__()
         self.lightning_trainer = lightning_trainer
+        self.trainer_config = lightning_trainer.trainer_config
+        self.training_config = lightning_trainer.training
+
+        # for logging
+        self.total_timer = Timer()
+        self.snapshot_iterations = len(lightning_trainer.data_module.val_loader)
+        self.snapshot_iterations //= self.training_config.batch_size
```
Comment (ananthsub) [inline mmf/trainers/lightning_core/loop_callback.py:28]: could you set these attributes afterwards in on_train_start or on_test_start? there you have access to the trainer, so you don't need to pass it as an input to the constructor here. otherwise you need to pass the trainer into the callback constructor and the callback into the trainer's constructor. not sure how that's working

File: mmf/trainers/lightning_core/loop_callback.py [hunk 2]
```diff
@@ -2,35 +2,196 @@
...
+        logger.info("Stepping into final validation check")
+        # Only do when run_type has train as it shouldn't happen on validation and
+        # inference runs. Inference will take care of this anyways. Also, don't run
+        # if current iteration is divisble by snapshot interval as it will just
+        # be a repeat
+        if (
+            "train" in self.lightning_trainer.run_type
+            and trainer.global_step % self.trainer_config.val_check_interval != 0
+        ):
+            self._start_eval(trainer)
```
Comment (ananthsub) [inline mmf/trainers/lightning_core/loop_callback.py:71]: i don't quite follow why evaluation needs to be started like this. why isn't the validation loop sufficient?

it seems like this callback is mixing both the logging of event times (could the profiler be used instead?) along with the evaluation hook. imo these should be separated

File: mmf/trainers/lightning_core/loop_callback.py [hunk 3]
```diff
@@ -2,35 +2,196 @@
...
+    def _start_eval(self, trainer):
+        trainer.test(
+            model=trainer.model,
+            test_dataloaders=self.lightning_trainer.data_module.val_loader,
+        )
+
+    def _save_checkpoint(self, trainer):
+        logger.info("Checkpoint time. Saving a checkpoint.")
+        return
+        # TODO: sash Needs implementation - next mvp
```
Comment (ananthsub) [inline mmf/trainers/lightning_core/loop_callback.py:148]: why doesn't the existing model checkpoint callback work?

**Review by ananthsub — COMMENTED** [review]
I don't quite understand the role of the loop callback as it seems to be:
- handling checkpointing
- Logging timings
- Starting evaluations

Why not add multiple callbacks and let them run in parallel?

**Comment by ytsheng** [comment]
Closing in favor of: https://github.com/facebookresearch/mmf/pull/768
